Update FileSystem::pathByAppendingComponents() to take in a std::span

cdumez opened the pull request:
#### 3d0598740a712c651f60ab388da20b90956fe366
<pre>
Update FileSystem::pathByAppendingComponents() to take in a std::span
<a href="https://bugs.webkit.org/show_bug.cgi?id=293099">https://bugs.webkit.org/show_bug.cgi?id=293099</a>

Reviewed by Darin Adler.

Update FileSystem::pathByAppendingComponents() to take in a std::span
instead of a `const Vector&amp;`. This avoids having to construct a Vector
at many of the call sites.

* Source/WTF/wtf/FileSystem.cpp:
(WTF::FileSystemImpl::pathByAppendingComponents):
* Source/WTF/wtf/FileSystem.h:
* Source/WTF/wtf/cocoa/FileSystemCocoa.mm:
(WTF::FileSystemImpl::createTemporaryFileInDirectory):
* Source/WTF/wtf/posix/FileSystemPOSIX.cpp:
(WTF::FileSystemImpl::pathByAppendingComponents):
* Source/WebCore/workers/service/server/SWScriptStorage.cpp:
(WebCore::SWScriptStorage::registrationDirectory const):
* Source/WebKit/NetworkProcess/storage/BackgroundFetchStoreManager.cpp:
(WebKit::BackgroundFetchStoreManager::clearFetch):
(WebKit::BackgroundFetchStoreManager::clearAllFetches):
(WebKit::BackgroundFetchStoreManager::storeFetchAfterQuotaCheck):
* Source/WebKit/NetworkProcess/storage/CacheStorageDiskStore.cpp:
(WebKit::recordFilePathWithDirectory):
* Source/WebKit/NetworkProcess/storage/CacheStorageManager.cpp:
(WebKit::CacheStorageManager::originsOfCacheStorageData):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::originDirectoryPath):
* Source/WebKit/UIProcess/API/glib/WebKitWebsiteDataManager.cpp:
(webkitWebsiteDataManagerGetFaviconDatabasePath):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerConfigurationCocoa.mm:
(WebKit::WebExtensionControllerConfiguration::createStorageDirectoryPath):
* Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
(WebKit::WebsiteDataStore::defaultCookieStorageFile):
* Source/WebKit/webpushd/WebPushDaemonMain.mm:
(WebKit::getWebPushDirectoryPathWithMigrationIfNecessary):
* Tools/TestWebKitAPI/Tests/WTF/FileSystem.cpp:
(TestWebKitAPI::TEST_F(FileSystemTest, deleteNonEmptyDirectory)):
(TestWebKitAPI::TEST_F(FileSystemTest, makeAllDirectories)):
(TestWebKitAPI::TEST_F(FileSystemTest, pathFileName)):
(TestWebKitAPI::TEST_F(FileSystemTest, parentPath)):
(TestWebKitAPI::TEST_F(FileSystemTest, pathByAppendingComponents)):
(TestWebKitAPI::TEST_F(FileSystemTest, listDirectory)):
(TestWebKitAPI::TEST_F(FileSystemTest, realPath)):
* Tools/TestWebKitAPI/Tests/WebCore/win/WebCoreBundle.cpp:
(TEST_F(WebCoreBundleTest, BundlePathFromPath)):
(TEST_F(WebCoreBundleTest, BundlePathFromNameTypeDirectory)):
(TEST_F(WebCoreBundleTest, BundlePathFromComponents)):

Canonical link: <a href="https://commits.webkit.org/295020@main">https://commits.webkit.org/295020@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3ca2d08d4b6c541670e57fc8fad4bbc2acadd6eb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/103816 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/23518 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/13838 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/109009 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/54468 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/105856 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/23871 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/32064 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/78878 "Build was cancelled. Recent messages:Printed configuration") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/106822 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/18543 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/93657 "Build was cancelled. Recent messages:OS: Sonoma (14.7.3), Xcode: 15.4; Skipping applying patch since patch_id isn't provided; Checked out pull request; run-api-tests (cancelled)") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/59212 "Build was cancelled. Recent messages:Printed configuration") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/18345 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/11706 "Passed tests") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/53844 "Build was cancelled. Recent messages:Printed configuration; Running apply-patch; Checked out pull request; Compiled WebKit (cancelled)") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/96491 "Build was cancelled. Recent messages:OS: Sonoma (14.7.3), Xcode: 15.4; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled JSC (warnings); jscore-tests (cancelled)") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/88105 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/11768 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/111396 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/102427 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/30972 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/22841 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/87884 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/31334 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/89858 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/87531 "Build was cancelled. Recent messages:Printed configuration; Running apply-patch; Checked out pull request; 101 api tests failed or timed out; 101 api tests failed or timed out") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22288 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/32423 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/10150 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/25336 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/30900 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/36203 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/126060 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/30694 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/34895 "Found 1 new JSC stress test failure: wasm.yaml/wasm/stress/cc-int-to-int-no-jit.js.default-wasm (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/34029 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/32255 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WTF/wtf/cocoa/FileSystemCocoa.mm:
```diff
@@ -189,7 +189,7 @@ String extractTemporaryZipArchive(const String& path)
 {
     auto fsSuffix = fileSystemRepresentation(suffix);
-    auto templatePath = pathByAppendingComponents(directory, { StringView { "XXXXXX"_s }, StringView { suffix } });
+    auto templatePath = pathByAppendingComponents(directory, span<StringView>({ "XXXXXX"_s, suffix }));
```
I guess this does not create temporary `String` objects? I couldn’t immediately tell by code inspection. Sometimes a reference is copied into an object, but maybe not when making an initializer list?

cdumez replied:
I have just tried the following in godbolt:
```
class MyString {
public:
    MyString(const char* s)
    : m_string(s) {
        printf("Constructing MyString\n");
    }

    MyString(const MyString& other)
        : m_string(other.m_string)
    {
        printf("Copying MyString\n");
    }

    MyString& operator=(const MyString& other)
    {
        printf("Copy assignment for MyString\n");
        m_string = other.m_string;
        return *this;
    }

    const std::string& stdString() const { return m_string; }
private:
    std::string m_string;
};

class MyStringView {
public:
    MyStringView(const MyString& s)
    : m_view(s.stdString()) { }

private:
    std::string_view m_view;
};

static void foo(std::span<const MyStringView> sp)
{
}

int main() {
    MyString s1("a");
    MyString s2("b");
    foo(std::initializer_list<MyStringView> { s1, s2 });
    return 0;
}
```

It prints:
```
Constructing MyString
Constructing MyString
```
So I don't think any copy of the String is happening here. Let me know if you notice something wrong with my testing.

darinadler commented on Source/WTF/wtf/FileSystem.h:
```diff
@@ -95,7 +95,7 @@ WTF_EXPORT_PRIVATE bool updateFileModificationTime(const String& path); // Sets
 WTF_EXPORT_PRIVATE bool isHiddenFile(const String&);
 WTF_EXPORT_PRIVATE String pathByAppendingComponent(StringView path, StringView component);
-WTF_EXPORT_PRIVATE String pathByAppendingComponents(StringView path, const Vector<StringView>& components);
+WTF_EXPORT_PRIVATE String pathByAppendingComponents(StringView path, std::span<const StringView> components);
```
It’s a little annoying that we have to write `span<StringView>` at each call site. Could we have this take `std::initializer_list<StringView>` instead, at least for now? Anything that would make all the call sites easier to read might be a good choice as long as there are no correctness problems. But maybe that’s not a good idea?

cdumez replied:
I can totally have an explicit `std::initializer_list<StringView> { }` at call sites. I actually introduced the `span()` overload to try and simplify these call sites. I admit, this is only a minor simplification when you have to specify the type instead the initializer list (StringView here).

However, for cases where the type can be deduced, I think `span({ 1, 2, 3})` simplifies things a lot compared to `std::initializer_list<int>({ 1, 2, 3})`. Sadly, just passing `{ 1, 2, 3 }` won't build until C++26.

If you prefer being explicit, about std::intializer_list though, I don't mind going back to that instead.

darinadler replied:
If there’s a way to make the call sites less wordy, we should do it. But if not, I understand. I don’t think we need to prefer `std::initializer_list()` over `span()`, but if we are using `span<StringView>` to convert types, then I think it’s not just a workaround. We’d still need it once the C++26 feature shows up, or we’d need to be explicit about `std::initializer_list` instead.

darinadler commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -105,7 +105,7 @@ static String originDirectoryPath(const String& rootPath, const WebCore::ClientO
     auto encodedTopOrigin = encode(origin.topOrigin.toString(), salt);
     auto encodedOpeningOrigin = encode(origin.clientOrigin.toString(), salt);
-    return FileSystem::pathByAppendingComponents(rootPath, { encodedTopOrigin, encodedOpeningOrigin });
+    return FileSystem::pathByAppendingComponents(rootPath, std::initializer_list<StringView> { encodedTopOrigin, encodedOpeningOrigin });
```
I don’t understand why in this case we need to write `std::initializer_list<StringView>` but in the other cases we write `span<StringView>`.

cdumez replied:
Oh, there is no need. We should be consistent one way or another.

cdumez committed:
Update FileSystem::pathByAppendingComponents() to take in a std::span
https://bugs.webkit.org/show_bug.cgi?id=293099

Reviewed by Darin Adler.

Update FileSystem::pathByAppendingComponents() to take in a std::span
instead of a `const Vector&`. This avoids having to construct a Vector
at many of the call sites.

* Source/WTF/wtf/FileSystem.cpp:
(WTF::FileSystemImpl::pathByAppendingComponents):
* Source/WTF/wtf/FileSystem.h:
* Source/WTF/wtf/cocoa/FileSystemCocoa.mm:
(WTF::FileSystemImpl::createTemporaryFileInDirectory):
* Source/WTF/wtf/posix/FileSystemPOSIX.cpp:
(WTF::FileSystemImpl::pathByAppendingComponents):
* Source/WebCore/workers/service/server/SWScriptStorage.cpp:
(WebCore::SWScriptStorage::registrationDirectory const):
* Source/WebKit/NetworkProcess/storage/BackgroundFetchStoreManager.cpp:
(WebKit::BackgroundFetchStoreManager::clearFetch):
(WebKit::BackgroundFetchStoreManager::clearAllFetches):
(WebKit::BackgroundFetchStoreManager::storeFetchAfterQuotaCheck):
* Source/WebKit/NetworkProcess/storage/CacheStorageDiskStore.cpp:
(WebKit::recordFilePathWithDirectory):
* Source/WebKit/NetworkProcess/storage/CacheStorageManager.cpp:
(WebKit::CacheStorageManager::originsOfCacheStorageData):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::originDirectoryPath):
* Source/WebKit/UIProcess/API/glib/WebKitWebsiteDataManager.cpp:
(webkitWebsiteDataManagerGetFaviconDatabasePath):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerConfigurationCocoa.mm:
(WebKit::WebExtensionControllerConfiguration::createStorageDirectoryPath):
* Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
(WebKit::WebsiteDataStore::defaultCookieStorageFile):
* Source/WebKit/webpushd/WebPushDaemonMain.mm:
(WebKit::getWebPushDirectoryPathWithMigrationIfNecessary):
* Tools/TestWebKitAPI/Tests/WTF/FileSystem.cpp:
(TestWebKitAPI::TEST_F(FileSystemTest, deleteNonEmptyDirectory)):
(TestWebKitAPI::TEST_F(FileSystemTest, makeAllDirectories)):
(TestWebKitAPI::TEST_F(FileSystemTest, pathFileName)):
(TestWebKitAPI::TEST_F(FileSystemTest, parentPath)):
(TestWebKitAPI::TEST_F(FileSystemTest, pathByAppendingComponents)):
(TestWebKitAPI::TEST_F(FileSystemTest, listDirectory)):
(TestWebKitAPI::TEST_F(FileSystemTest, realPath)):
* Tools/TestWebKitAPI/Tests/WebCore/win/WebCoreBundle.cpp:
(TEST_F(WebCoreBundleTest, BundlePathFromPath)):
(TEST_F(WebCoreBundleTest, BundlePathFromNameTypeDirectory)):
(TEST_F(WebCoreBundleTest, BundlePathFromComponents)):

Canonical link: https://commits.webkit.org/295020@main
